SEP-34: Wallet Attribution (#709)
stellar/stellar-protocol: ecosystem/sep-0034.md

Author: Jake Urban, Leigh McCulloch
Track: Standard
Status: Draft
Created: 2020-08-17
Discussion: https://github.com/stellar/stellar-protocol/issues/685
```

## Simple Summary
This protocol defines how anchors can cryptographically attribute HTTPS requests to wallet applications. 

## Motivation
In general, web services may want to attribute resources created or actions taken on their service to client applications. Because requests headers like `User-Agent` can be spoofed, these web services require a cryptographically secure method of verifying a client's identity.

Throughout this document, the resource mentioned is a deposit or withdrawal transaction, but the resource or action that requires identity verification could be related to different use cases between wallets and anchors.

## Abstract
This following protocol involves 3 parties:

- The Wallet: the application that makes HTTPS requests to the anchor server for a resource or action
- The Wallet Server: the web server that claims ownership of the wallet
    - Note that frontend-only wallets _must_ add a backend web server to comply with this protocol
- The Anchor Server: the web server that verifies the wallet's claims

> At a high level, the Wallet Server generates a [JSON Web Signature](https://www.rfc-editor.org/rfc/rfc7515.html) (JWS) that the Wallet receives and includes in future requests to the Anchor Server. The Anchor Server verifies that the JWS included in the request was signed by the Wallet Server and that the transaction (or resource) ID was included in the decoded [JWT](https://tools.ietf.org/html/rfc7519) payload. These steps prove that the Wallet Server takes ownership of the requests being made for a particular transaction.

Reviewer: could JWT be defined as JWS is, i.e. JSON Web Token (JWT)

Author: Yes, good idea

## Preamble

```
SEP: 0034
Title: Wallet Request Attribution for Anchors
Author: Jake Urban, Leigh McCulloch
Track: Standard
Status: Draft
Created: 2020-08-17
Discussion: https://github.com/stellar/stellar-protocol/issues/685
```

## Simple Summary
This protocol defines how anchors can cryptographically attribute transaction requests to wallet applications. 

## Motivation

> In general, web services may want to attribute resources created or actions taken on their service to client applications. Because requests headers like `User-Agent` can be spoofed, anchors require a cryptographically secure method of verifying a client's identity.

Author: I think the majority of the document is specific to wallets and anchors, but I'd like to keep at least one mention of the fact that this _could_ be generalized. I don't feel too strongly on this though.

> In general, web services may want to attribute resources created or actions taken on their service to client applications. Because requests headers like `User-Agent` can be spoofed, anchors require a cryptographically secure method of verifying a client's identity.

Author: I've updated the section to better explain the SEP's main use case but also leave room for generalization.

> For example, an anchor server may not charge fees for transactions created using a particular wallet. In this case, the resource is the deposit or withdrawal transaction, but the resource or action could be related to different use cases.

Author: If you're of the opinon that there should be no mention of other use cases thats fine, but I image there will be a future conversation about generalizing this SEP for a different identity verification use case.

> For example, an anchor server may not charge fees for transactions created using a particular wallet. In this case, the resource is the deposit or withdrawal transaction, but the resource or action could be related to different use cases.

Author: I've updated this section, let me know if you think it still needs work.

## Abstract
This following protocol involves 3 parties:

> - the wallet: the application requesting the transaction with the anchor server 

Author: This is not helpful, because this JWS token isn't used for all requests like the SEP-10 token. I vote to keep this the way it is.

- the wallet signing server: the web server that claims ownership of the wallet

> - the anchor server: the web server that verifies the signing server's claims

Reviewer 2: ```suggestion
- The Anchor Server: the web server that verifies the wallet's claims
```

> At a high level, the signing server generates a [JSON Web Signature](https://www.rfc-editor.org/rfc/rfc7515.html) (JWS) that the wallet application receives and includes in future requests to the anchor server. The anchor server verifies that the token included in the request was signed by the signing server, proving that the signing server takes ownership for the client requesting the transaction with the anchor server.

Reviewer 2: ```suggestion
At a high level, the wallet server generates a [JSON Web Signature](https://www.rfc-editor.org/rfc/rfc7515.html) (JWS) that the wallet application receives and includes in future requests to the anchor server. The anchor server verifies that the token included in the request was signed by the wallet server, proving that the wallet takes ownership of the request being made.
```

> How the client application communicates with the signing server and receives the JWS is out of the scope of this document. The only requirement is for the signing server to return a JWS in the format descibed in the next section. Wallet signing servers should not accept requests for tokens if the client cannot be authenticated or if the request omits the transaction ID the wallet received from the anchor.

Reviewer 2: ```suggestion
How the wallet communicates with it's server and receives the JWS is out of the scope of this document. The only requirement is for the wallet server to return a JWS in the format described in the next section. Wallet servers should not accept requests for tokens if the client cannot be authenticated or if the request omits the transaction ID the wallet received from the anchor.
```

## Token

> The JWS can be generated using two JSON objects and a JSON Web Key ([JWK](https://www.rfc-editor.org/rfc/rfc7517.html)). The JWK itself is generated using the wallet signing server's `SIGNING_KEY` and associated secret key listed in the server's [`stellar.toml`](sep-0001.md).

Reviewer 2: ```suggestion
The JWS is generated using two JSON objects and a JSON Web Key ([JWK](https://www.rfc-editor.org/rfc/rfc7517.html)). The JWK itself is generated using the wallet signing server's `SIGNING_KEY` and associated secret key listed in the server's [`stellar.toml`](sep-0001.md).
```

    "exp": "1597789801",
    "iat": "1597703375"
}
```

* `iss` — the fully qualified URL of the home domain for the signing server issuing the token
* `sub` - the stellar account address of the user requesting the transaction
* `jti` — the identifier for the transaction being processed by the client and defined by the anchor
* `kid` - the stellar public key address used to sign the JWS
* `aud` - the fully qualified URL of the home domain for the anchor server receiving the token
* `iat` - the [timestamp](https://tools.ietf.org/html/rfc7519#section-4.1.6) for the time this token was issued
* `exp` - the [timestamp](https://tools.ietf.org/html/rfc7519#section-4.1.4) for the time on or after which the JWT must not be accepted for processing

The two JSON objects should be base64-encoded and concatenated with a `.` character. 
```
eyJhbGciOiJFZERTQSJ9
.
eyJhdWQiOiJodHRwczovL2FuY2hvcnNlcnZlci5jb20iLCJleHAiOiIxNTk3Nzg5ODAxIiwiaWF0IjoiMTU5NzcwMzM3NSIsImlzcyI6Imh0dHBzOi8vd2FsbGV0c2lnbmluZ3NlcnZlci5jb20iLCJqdGkiOiJhYTc3OTgzYS1lNTUwLTRkOTAtOGNjMi1kNjYxZDdmMGI4ZjYiLCJzdWIiOiJHQ0NYS0NIVkVESDZNR1pZSUhER0lGRDVMSEdEM1JDQVo3TkZMSUFZNk5aQ0ZZQlZTNk1RUEZSNyJ9
```

> The resulting string should then be signed by the server's JSON Web Key (JWK). JWK's can be constructed by using the Stellar public and private keys used by client.

Reviewer 2: I think the following code, both the js portion and the python are a bit confusing. 

I suggest not having any code here, just a step by step example process to getting from a stellar keypair to a signed request. Also, if we avoid using random keypairs we can remove statements of the "this *may* look something like" kind.

Once we have SDK support we can retrofit this SEP with proper code examples

Author: How would you describe the process of decomposing a stellar keypair into the raw bytes and into the base64 representation? The only way I know how to communicate that process is via an SDK example.

I agree I'd like to not have code in the SEP, but some of these processes are really hard to explain without examples, like how to sign a JWS with a JWK. How do I give steps for this without getting into the RFC details?

        "iat": "1597703375"
    }   
)

# The base64-encoded public and secret keys
# Stellar uses its own representation, so we use the raw bytes
public_key_encoded = codecs.encode(kp.raw_public_key(), "base64")
secret_key_encoded = codecs.encode(kp.raw_secret_key(), "base64")

# the JWK used to sign the token
key = jwk.JWK(
    kty="OKP",
    crv="Ed25519",
    x=public_key_encoded,
    d=secret_key_encoded,
)
jws.make_signed_token(key)
print(jws.serialize())
```

The resulting string will look something like this:
```
eyJhbGciOiJFZERTQSJ9
.
eyJhdWQiOiJodHRwczovL2FuY2hvcnNlcnZlci5jb20iLCJleHAiOiIxNTk3Nzg5ODAxIiwiaWF0IjoiMTU5NzcwMzM3NSIsImlzcyI6Imh0dHBzOi8vd2FsbGV0c2lnbmluZ3NlcnZlci5jb20iLCJqdGkiOiJhYTc3OTgzYS1lNTUwLTRkOTAtOGNjMi1kNjYxZDdmMGI4ZjYiLCJzdWIiOiJHQ0NYS0NIVkVESDZNR1pZSUhER0lGRDVMSEdEM1JDQVo3TkZMSUFZNk5aQ0ZZQlZTNk1RUEZSNyJ9
.
xZi9TX_KI9sXA1XJlm_O5t6FLxs0sx9suBf5Kv4SVqL8Glo_zL4EbPZ6SgjeqBd31xnVQ-b5q4rwpiPIz278AQ
```

The public key used to generate the third part of the JWS can also be used to verify the integrity of the first and second part. This is what the anchor server would do when a wallet makes a request containing the JWS token. 

## Verifying Requests

> Other proposals will specify how the JWS returned by the client's signing server should be sent to the attribution server. The anchor can keep the public keys of known wallet signing servers, but if the client is unknown, the anchor should request the signing server's [stellar.toml](sep-0001.md) and validate that the value found was used to encrypt the JWS.

Reviewer 2: ```suggestion
Other proposals will specify how the JWS returned by the Wallet Server should be sent to the attribution server. The anchor can keep the public keys of known wallet  servers, but if the client is unknown, the anchor should request the wallet server's [stellar.toml](sep-0001.md) and validate that the value found was used to encrypt the JWS.
```

Wallet signing servers could also rotate their `SIGNING_KEY`. Anchors receiving JWS tokens from these clients may have to update their local copies of the wallets' public keys.

Once verified, the anchor server can use the JWT attributes with confidence that the integrity of the data has been maintained and that the wallet requesting the transaction and passing the JWS belongs to the wallet's signing server.

### Expiration

> Signing servers should select an expiration time for the JWT that is appropriate for the assumptions and risk of the interactions a user can perform with it. A client may own a transaction at the time the JWT is issued but they may not at some point in the future. 

Reviewer 2: ```suggestion
Wallet servers should select an expiration time for the JWT that is appropriate for the assumptions and risk of the interactions a user can perform with it. A wallet may own a request at the time the JWT is issued but they may not at some point in the future. 
```

Author: This makes it seem like the `request` is not an HTTP request, but a general request for a transaction. This should be disambiguated.

> Expiration times that are too long increase the risk that control of the transaction request will change. Expiration times that are too short increase the number of times authentication must reoccur, and a user using a hardware signing device or who must complete a complex signing process could have a poor user experience.

Reviewer 2: Why is the user signing device relevant here?

Author: Good point its not. This is copied from SEP-10

## Preamble

```
SEP: To Be Assigned

> Title: Client Application Attribution

Author: I'll take up your suggestions to make the phrasing more concrete. We don't have another use case at the moment.

Author: Jake Urban, Leigh McCulloch
Track: Standard
Status: Draft
Created: 2020-08-17
Discussion: https://github.com/stellar/stellar-protocol/issues/685
```

## Simple Summary
This protocol outlines how a client can prove to a web server that the application requesting the relevant operation is owned by the client.

## Motivation
In general, web services may want to attribute incoming requests to client applications on a per-operation basis. Because requests headers like `User-Agent` can be spoofed, anchors require a cryptographically secure method of verifying a client's identity.

For example, an anchor server may not charge fees for transactions created using a particular wallet. In this case, the operation is processing a deposit or withdrawal transaction, but the operation could be any action related to different use cases.

## Abstract
This following protocol involves 3 parties:

- the client application: the party requesting an operation from the attribution server 

> - the client's signing server: the web server that claims ownership of the client application

Reviewer 2: I think that Morley's question is indicative that it's not very clear that `the client application` and `the client's signing server` are both operated by the same entity. 
Maybe `Wallet` and `Wallet Backend`?

> - the attribution server: the web server that verifies the signing server's claims

Reviewer 2: This is the Anchor Service, right?

At a high level, the signing server generates a [JSON Web Signature](https://www.rfc-editor.org/rfc/rfc7515.html) (JWS) that the client receives and includes in future requests to the attribution server. The attribution server verifies that the token included in the request was signed by the signing server, proving that the signing server takes ownership for the client requesting the operation from the verfication server.

> How the client application communicates with the signing server and receives the JWS is out of the scope of this document. The only requirement is for the signing server to return a JWS in the format descibed in the next section.

Reviewer 3: Why doesn't the SEP define this? If each anchor has their own API for accepting this JWS, it's going to be hard for clients to implement this feature.

Author: The client communicates with its own server to obtain the JWS, not the anchor it is trying to verify its identity with. We don't need to define the protocol between the client application and the client's signing server as long as the JWS is in the correct format. 

We could describe what a flow between the client application and signing server could look like, but it is useless to define a standard and expect it to be used since there is no need for interoperability.

## Token

The JWS can be generated using two JSON objects and a secret if the HMAC SHA-256 algorithm is used.

The first JSON object is the [JWS Protected Header](https://www.rfc-editor.org/rfc/rfc7515.html#section-4):
```json
{
    "typ": "JWT",

>     "alg": "HS256",

Reviewer 2: let's stick to Ed25519 so that we don't add new crypto algorithms to our sdks.

* `iat` (the time at which the JSW was issued [RFC7519, Section 4.1.6](https://tools.ietf.org/html/rfc7519#section-4.1.6)) — current timestamp (`1530644093`)
* `exp` (the expiration time on or after which the JWT must not be accepted for processing, [RFC7519, Section 4.1.4](https://tools.ietf.org/html/rfc7519#section-4.1.4)) — a server can pick its own expiration period for the token (`1530730493`)

Finally, the two JSON objects should be base64-encoded and concatenated with a `.` character. The resulting string should then be encrypted using a signing key, and the encrypted string should
be appended to the end of the unencrypted string, again including a `.` character.
```
eyJ0eXAiOiAiSldUIiwiYWxnIjogIkhTMjU2In0
.
eyJpc3MiOiAiaHR0cHM6Ly9hbmNob3Jkb21haW4uY29tL2F0dHJpYnV0aW9uIiwic3ViIjogImFhNzc5ODNhLWU1NTAtNGQ5MC04Y2MyLWQ2NjFkN2YwYjhmNiIsImV4cCI6ICIxNTk3Nzg5ODAxIiwiaWF0IjogIjE1OTc3MDMzNzUifQ
.
mEUU0apnMrlwiNER-Wac6F5xuXDWuZjAsmqy7Ap02C4
```

The signing key used to generate the third part of the JWS can also be used to verify it. 

## Verifying Requests

> Other proposals will specify how the JWS returned by the client's signing server should be sent to the attribution server. Once sent, the attribution server should check the signing server's [stellar.toml](sep-0001.md) for a `SIGNING_KEY` variable and validate that the value found was used to encrypt the JWS. 

Reviewer 2: We should link to concrete proposals once these are live
